File: 300 Longest Increasing Subsequence/300.java
import java.util.Arrays;

class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0; // corner case
        int max = 0;
        int[] cache = new int[n];
        Arrays.fill(cache, 1);
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if(nums[i] > nums[j]) {
                    cache[i] = Math.max(cache[i], cache[j] + 1);
                }
            }
            max = Math.max(max, cache[i]);
        }
        return max;
    }

    // Time Complexity: O(n^2)
    // Space Complexity: O(n)

    public int lengthOfLIS2(int[] nums) {
        int[] dp = new int[nums.length];
        int len = 0;
        for (int x : nums) {
            // insertion point that the index of the first element in the range greater than the key
            int i = Arrays.binarySearch(dp, 0, len, x);
            if (i < 0) i = -(i + 1);
            dp[i] = x;
            if (i == len) len++;
        }
        return len;
    }

    // Time Complexity: O(n * log(n))
    // Space Complexity: O(n)
}

class Solution {
    public static int findPositionToReplace(int[] a, int low, int high, int x) {
        int mid;
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (a[mid] == x)
                return mid;
            else if (a[mid] > x)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return low;
    }

    public static int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        int n = nums.length, len = 0;
        int[] increasingSequence = new int[n];
        increasingSequence[len++] = nums[0];
        for (int i = 1; i < n; i++) {
            if (nums[i] > increasingSequence[len - 1])
                increasingSequence[len++] = nums[i];
            else {
                int position = findPositionToReplace(increasingSequence, 0, len - 1, nums[i]);
                increasingSequence[position] = nums[i];
            }
        }
        return len;
    }
}

File: 300 Longest Increasing Subsequence/300.py
class Solution:
    def findPos(self, array, low, high, value):
        mid = 0
        if not array:
            return 0
        while low <= high:
            mid = (high + low) // 2
            if(array[mid] == value):
                return mid
            elif array[mid] < value:
                low = mid + 1
            else:
                high = mid - 1
        if low < len(array) and array[low] > value:
            return low
        elif high >= 0 and array[high] > value:
            return high
        return mid

    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        increseSequence = [nums[0]]
        len = 1
        for num in nums:
            if num > increseSequence[-1]:
                increseSequence.append(num)
                len += 1
            else:
                pos = self.findPos(increseSequence, 0, len - 1, num)
                increseSequence[pos] = num
        return len


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        stack = [nums[0]]
        for n in nums[1:]:
            if n > stack[-1]:
                stack.append(n)
            else:
                # binary search
                index = bisect.bisect_left(stack, n) 
                stack[index] = n
        return len(stack)